src/gbinder_fmq.c

```diff
@@ -0,0 +1,794 @@
+            grantor_pos++) {
+        GBinderFmqGrantorDescriptor *grantor = &grantors[grantor_pos];
+        guint32 grantor_fd_index;
+        gsize grantor_offset;
+        if (grantor_pos == DATA_PTR_POS && num_fds == 2) {
+            grantor_fd_index = 1;
+            grantor_offset = 0;
+        } else {
+            grantor_fd_index = 0;
+            grantor_offset = offset;
+            offset += mem_sizes[grantor_pos];
+        }
+        grantor->fd_index = grantor_fd_index;
+        grantor->offset = (guint32)(G_ALIGN8(grantor_offset));
+        grantor->extent = mem_sizes[grantor_pos];
+    }
+    return grantors;
+}
+
+void*
+gbinder_fmq_map_grantor_descriptor(
+    GBinderFmq* self,
+    guint32 index)
+{
+    const GBinderFds* fds = self->desc->data.fds;
+    gint fd_index;
+    gint map_offset;
+    gint map_length;
+    void* address;
+    void* ptr = NULL;
+
+    if (index >= self->desc->grantors.count) {
+        GWARN("grantor index must be less than %d", self->desc->grantors.count);
+    } else {
+        fd_index = gbinder_fmq_get_grantor_descriptor(self, index)->fd_index;
+        /* Offset for mmap must be a multiple of PAGE_SIZE */
+        map_offset = (gbinder_fmq_get_grantor_descriptor(self, index)->offset /
+            getpagesize()) * getpagesize();
+        map_length = gbinder_fmq_get_grantor_descriptor(self, index)->offset -
+            map_offset +
+            gbinder_fmq_get_grantor_descriptor(self, index)->extent;
+
+        address = mmap(0, map_length, PROT_READ | PROT_WRITE, MAP_SHARED,
+            gbinder_fds_get_fd(fds, fd_index), map_offset);
+        if (address == MAP_FAILED) {
+            GWARN("mmap failed: %d", errno);
+        } else {
+            ptr = (guint8*)(address) +
+                (gbinder_fmq_get_grantor_descriptor(self, index)->offset -
+                map_offset);
+        }
+    }
+    return ptr;
+}
+
+void
+gbinder_fmq_unmap_grantor_descriptor(
+    GBinderFmq* self,
+    void* address,
+    guint32 index)
+{
+    gint map_offset;
+    gint map_length;
+    void* base_address;
+
+    if (index >= self->desc->grantors.count) {
+        GWARN("grantor index must be less than %d", self->desc->grantors.count);
+    } else {
+        map_offset = (gbinder_fmq_get_grantor_descriptor(self, index)->offset /
+            getpagesize()) * getpagesize();
+        map_length =
+            gbinder_fmq_get_grantor_descriptor(self, index)->offset - map_offset
+            + gbinder_fmq_get_grantor_descriptor(self, index)->extent;
+
+        base_address = (guint8*)(address) -
+            (gbinder_fmq_get_grantor_descriptor(self, index)->offset -
+            map_offset);
+        if (base_address) {
+            munmap(base_address, map_length);
+        }
+    }
+}
+
+void
+gbinder_fmq_free(
+    GBinderFmq* self)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+
+    if (self->desc->flags == GBINDER_FMQ_TYPE_UNSYNC_WRITE &&
+            priv->read_ptr != NULL) {
```

Comment:
There's no point in this `priv->read_ptr != NULL` check because `g_free` is NULL tolerant (i.e. will check it again) and the the `else` branch makes that check again. ETOOMANYCHECKS

Reply:
Removed useless checks.